### PR TITLE
[MAINTENANCE] Batch ID must incorporate batch_spec_passthrough.  Instantiate Validator with DataContext in test modules.  Query metrics/expectations types cleanup.

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* `Batch` ID must incorporate `batch_spec_passthrough` (when non-NULL).
* Creating `Validator` in test modules must utilize `DataContext` (otherwise, it is non-robust for subsequent processing).
* Clean up `Query` metrics and expectations to utilize proper return types.
* Update tests.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1358/GREAT-1479
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/expectations/metrics/query_metrics/query_column.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def _sqlalchemy(
         metric_value_kwargs: dict,
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/query_metrics/query_column.py:35]: This file had `7` typing errors

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L100

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def _sqlalchemy(
         metric_value_kwargs: dict,
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/query_metrics/query_column_pair.py:36]: This file had 9 typing errors

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L99

**Review by Kilo59 — COMMENTED** [review]
Several of these metrics modules have a low number of typing errors.
Can you remove at least one of them from `pyproject.toml` exclude to verify the safety of these changes?

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L98-L101

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py [hunk 3]
```diff
@@ -73,9 +69,12 @@ def _validate(
     ) -> Union[ExpectationValidationResult, dict]:
         diff: Union[float, int] = configuration["kwargs"].get("diff")
         mostly: str = configuration["kwargs"].get("mostly")
-        query_result: Union[sqlalchemy_engine_Row, pyspark_sql_Row] = metrics.get(
-            "query.column_pair"
-        )
+        query_result = metrics.get("query.column_pair")
```
Comment (Shinnnyshinshin) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py:72]: question : why don't we do the same `convert_to_json_serializable` step here?

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_be_unique_with_condition.py [hunk 4]
```diff
@@ -70,9 +70,8 @@ def _validate(
         runtime_configuration: dict = None,
         execution_engine: ExecutionEngine = None,
     ) -> Union[ExpectationValidationResult, dict]:
-
         query_result = metrics.get("query.template_values")
```
Comment (Shinnnyshinshin) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_be_unique_with_condition.py:71]: same question

File: tests/datasource/test_new_datasource_with_sql_data_connector.py [hunk 5]
```diff
@@ -891,7 +891,7 @@ def test_batch_request_sql_with_schema(
         "runtime_parameters": {"query": "SELECT * FROM main.table_2"},
         "batch_identifiers": {
             "pipeline_stage_name": "core_processing",
-            "airflow_run_id": 1234567890,
+            "airflow_run_id": 1234567891,
```
Comment (Shinnnyshinshin) [inline tests/datasource/test_new_datasource_with_sql_data_connector.py:894]: Non-blocking, just curious : why the `0` -> `1`?

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py [hunk 6]
```diff
@@ -36,7 +36,7 @@ def _sqlalchemy(
...
         metrics: Dict[str, Any],
         runtime_configuration: dict,
-    ) -> List[sqlalchemy_engine_Row]:
+    ) -> List[dict]:
```
Comment (Shinnnyshinshin) [inline great_expectations/expectations/metrics/query_metrics/query_column_pair.py:39]: What is the reason for this change?

File: tests/datasource/data_connector/test_sql_data_connector.py [hunk 7]
```diff
@@ -2453,4 +2551,4 @@ def test_include_schema_name_get_available_data_assets(
 
     expected_data_asset_names = ["test_ci.test_df", "test_connection.test_df"]
 
-    assert actual_data_asset_names == expected_data_asset_names
+    assert set(expected_data_asset_names).issubset(set(actual_data_asset_names))
```
Comment (Shinnnyshinshin) [inline tests/datasource/data_connector/test_sql_data_connector.py:2539]: :heart:

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_query_expectations.md [hunk 8]
```diff
@@ -224,7 +224,7 @@ This:
 
 Becomes something like this:
 
-```python file=../../../../tests/integration/docusaurus/expectations/creating_custom_expectations/expect_queried_table_row_count_to_be.py#L30-L33
```
Comment (Shinnnyshinshin) [inline docs/guides/expectations/creating_custom_expectations/how_to_create_custom_query_expectations.md:227]: Thank you for catching these. One request to double-check these are still accurate after review / possible revisions :)

File: contrib/experimental/great_expectations_experimental/expectations/expect_query_to_have_no_duplicate_value_combinations.py [hunk 9]
```diff
@@ -56,18 +52,20 @@ def _validate(
...
-        )
-        success: bool = query_result == []
         columns = configuration["kwargs"].get("columns")
         duplicates = [
             dict(zip(columns + ["no_occurrences"], row)) for row in query_result
         ]
 
         return {
-            "success": success,
+            "success": not query_result,
```
Comment (Shinnnyshinshin) [inline contrib/experimental/great_expectations_experimental/expectations/expect_query_to_have_no_duplicate_value_combinations.py:64]: :heart:

**Review by Shinnnyshinshin — COMMENTED** [review]
Thank you very much for this Alex. Left a few comments/ questions

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py (hunk 3 above)
Comment (alexsherstinsky) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py:72]: @Shinnnyshinshin It is not always needed, so reducing the need for this relatively slow serialization step whenever possible!  Thanks. :)

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_be_unique_with_condition.py (hunk 4 above)
Comment (alexsherstinsky) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_be_unique_with_condition.py:71]: @Shinnnyshinshin Thank you -- just trying to use it only when absolutely necessary.

File: tests/datasource/test_new_datasource_with_sql_data_connector.py (hunk 5 above)
Comment (alexsherstinsky) [reply tests/datasource/test_new_datasource_with_sql_data_connector.py:894]: @Shinnnyshinshin This is critical -- thank you for drawing attention to this.  We must disambiguate `Batch` ID -- otherwise, multi-batch scenarios will fail (batch metric service will see that the value exists and return it, instead of computing new one).  So this is a future-proofing hygiene step.

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/query_metrics/query_column_pair.py:39]: @Shinnnyshinshin This is also very important.  The job of the metric is to return backend-agnostic output, regardless of the backend.  In the map cases, the decorator will take care of it, but here we are dealing with a raw metric, so it must take care of it.

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_query_expectations.md (hunk 8 above)
Comment (alexsherstinsky) [reply docs/guides/expectations/creating_custom_expectations/how_to_create_custom_query_expectations.md:227]: @Shinnnyshinshin This was a fix to reflect my own changes -- it took me a while to realize what was going on, but then I was happy to be able to draw upon our erstwhile experience with Docusaurus and fix it.  So I did the double checking already by virtue of fixing the line numbers.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py [hunk 10]
```diff
@@ -66,9 +67,8 @@ def _validate(
         runtime_configuration: dict = None,
         execution_engine: ExecutionEngine = None,
     ) -> Union[ExpectationValidationResult, dict]:
-
-        query_result = metrics.get("query.template_values")
-        num_of_duplicates = query_result[0][0]
+        metrics = convert_to_json_serializable(data=metrics)
```
Comment (billdirks) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py:68]: Why do we need to make sure the dict is json serializable here since we never serialize it? If you need to convert this 1 value, can we grab it and then just convert this value?

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py (hunk 3 above)
Comment (billdirks) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py:72]: What is the type of query result here? Since we iterate through it and then make it a dict I'm wondering if we can do `query_result = dict(metrics.get("query.column_pair"), [])`.

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py (hunk 6 above)
Comment (billdirks) [reply great_expectations/expectations/metrics/query_metrics/query_column_pair.py:39]: Do you anything about the structure of the `dict`? Are the keys `str`s or is this the result of processing data that can come from anywhere and we don't know the structure?

File: tests/datasource/data_connector/test_sql_data_connector.py [hunk 11]
```diff
@@ -578,11 +579,34 @@ def test_example_H(test_cases_for_sql_data_connector_sqlite_execution_engine):
...
 @pytest.mark.parametrize("sampler_method_name_prefix", ["_", ""])
 def test_get_batch_data_and_markers_sampling_method__limit(
+    in_memory_runtime_context,
     sampler_method_name_prefix,
+    test_cases_for_sql_data_connector_sqlite_connection_url,
     test_cases_for_sql_data_connector_sqlite_execution_engine,
 ):
     execution_engine = test_cases_for_sql_data_connector_sqlite_execution_engine
 
+    in_memory_runtime_context.datasources["my_test_datasource"] = Datasource(
```
Comment (billdirks) [inline tests/datasource/data_connector/test_sql_data_connector.py]: I like the use of in memory contexts! It looks like this Datasource is used in a few places. Could you put it in some utility method or fixture so we don't need to copy and paste it?

File: great_expectations/expectations/metrics/query_metrics/query_column.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/query_metrics/query_column.py:35]: Taken care of.

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/query_metrics/query_column_pair.py:36]: Taken care of.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this. I left a few questions inline.

File: tests/datasource/data_connector/test_sql_data_connector.py (hunk 11 above)
Comment (alexsherstinsky) [reply tests/datasource/data_connector/test_sql_data_connector.py]: @billdirks I can try -- it may not always be the same, but I think some repetition savings can be accomplished.  Back soon.

File: pyproject.toml [hunk 12]
```diff
@@ -96,8 +96,6 @@ exclude = [
     'expectations/metrics/column_aggregate_metrics/column_', # 21
     'expectations/metrics/map_metric_provider\.py',  # 57
     'expectations/metrics/metric_provider\.py',  # 12
-    'expectations/metrics/query_metrics/query_column_pair\.py', # 9
-    'expectations/metrics/query_metrics/query_column\.py', # 7
```
Comment (Kilo59) [inline pyproject.toml:100]: 🚀

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py (hunk 10 above)
Comment (alexsherstinsky) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py:68]: @billdirks This can come from profiler, which will return `Numpy` array.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_have_n_distinct_values_with_condition.py [hunk 13]
```diff
@@ -70,7 +70,6 @@ def _validate(
         runtime_configuration: dict = None,
         execution_engine: ExecutionEngine = None,
     ) -> Union[ExpectationValidationResult, dict]:
-
```
Comment (Kilo59) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_have_n_distinct_values_with_condition.py:73]: You have a whitespace change here.

**Review by Kilo59 — APPROVED** [review]
Pending resolution of every one else requests/suggestions ✔️

File: tests/datasource/data_connector/test_sql_data_connector.py (hunk 11 above)
Comment (alexsherstinsky) [reply tests/datasource/data_connector/test_sql_data_connector.py]: @billdirks Done.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_have_n_distinct_values_with_condition.py (hunk 13 above)
Comment (alexsherstinsky) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_to_have_n_distinct_values_with_condition.py:73]: @Kilo59 Yes -- I deleted it for better aesthetics (do we need it?).

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py (hunk 10 above)
Comment (billdirks) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_list_to_be_unique.py:68]: Does that mean `metrics` doesn't have a consistent interface unless you do this conversion?

File: tests/datasource/data_connector/test_configured_asset_glue_catalog_data_connector.py [hunk 14]
```diff
@@ -762,441 +769,546 @@ def test_get_batch_data_and_metadata_with_sampling_method__random_in_batch_spec_
...
-    my_data_connector = ConfiguredAssetAWSGlueDataCatalogDataConnector(
-        name="my_glue_catalog_data_connector",
-        datasource_name="FAKE_Datasource_NAME",
-        execution_engine=execution_engine,
-        assets={
-            "db_test.tb_titanic": {
-                "table_name": "tb_titanic_with_partitions",
-                "database_name": "db_test",
-            }
+    in_memory_runtime_context.datasources["FAKE_Datasource_NAME"] = Datasource(
```
Comment (billdirks) [inline tests/datasource/data_connector/test_configured_asset_glue_catalog_data_connector.py:779]: Oh, I hadn't opened up the large diff file. This looks repeated, could you make a util method or fixture for this and the `BatchDefinition` below?

**Review by Shinnnyshinshin — APPROVED** [review]
LGTM pending @billdirks's questions as well.

**Review by billdirks — APPROVED** [review]

**Comment by itaise** [comment]
@alexsherstinsky 
We tried to upgrade to new great expectations version, and a test that I contributed is no longer working :disappointed: we see it is because of this PR
this is the test that no longer work:
[expect_queried_slowly_changing_table_to_have_no_gaps](https://github.com/great-expectations/great_expectations/commit/5d936aaa77b0892832e1342fd0f04af7504bb371#diff-4069b72cc6fc4b0b141c80690594480a041965c753a2620279a0701060e5ba20)
can you please check? thanks a lot and happy holidays!
https://github.com/great-expectations/great_expectations/commit/5d936aaa77b0892832e1342fd0f04af7504bb371

**Comment by alexsherstinsky** [comment]
> @alexsherstinsky We tried to upgrade to new great expectations version, and a test that I contributed is no longer working 😞 we see it is because of this PR this is the test that no longer work: [expect_queried_slowly_changing_table_to_have_no_gaps](https://github.com/great-expectations/great_expectations/commit/5d936aaa77b0892832e1342fd0f04af7504bb371#diff-4069b72cc6fc4b0b141c80690594480a041965c753a2620279a0701060e5ba20) can you please check? thanks a lot and happy holidays! [5d936aa](https://github.com/great-expectations/great_expectations/commit/5d936aaa77b0892832e1342fd0f04af7504bb371)

Hi @itaise -- Sorry that this expectation is not working (which is surprising, because all tests passed).  In any case, I am looking at it right now.  If you would like more frequent status updates, please reach out to me in the Great Expectations Community Slack.  Thank you very much!

**Comment by alexsherstinsky** [comment]
@itaise The fix appears in #6650 -- thank you!
